From f142ad23506f79e3210d78608f31a0bb4d16d154 Mon Sep 17 00:00:00 2001
From: Gustavo Duarte <gduarte@redhat.com>
Date: Tue, 18 Nov 2014 14:08:17 -0500
Subject: [powerpc] use device_online/offline() instead of cpu_up/down()

Message-id: <1416319697-20018-1-git-send-email-gduarte@redhat.com>
Patchwork-id: 100188
O-Subject: [RHEL7.1 PATCH BZ 1157737] powerpc: use device_online/offline() instead of cpu_up/down()
Bugzilla: 1157737
RH-Acked-by: David Gibson <dgibson@redhat.com>
RH-Acked-by: Oleg Nesterov <oleg@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>

BZ 1157737
Upstream Status: 10ccaf178b2b961d8bca252d647ed7ed8aae2a20
Build Info: http://brewweb.devel.redhat.com/brew/taskinfo?taskID=8255834
ppc64le: http://brewweb.devel.redhat.com/brew/taskinfo?taskID=8255846
Tested: IBM verified it fixes the issue. I also applied on top of -204 and ran
cpu hotplug sanity checks on ppc64 and ppc64le. Everything is working fine.

commit 10ccaf178b2b961d8bca252d647ed7ed8aae2a20
Author: Dan Streetman <ddstreet@ieee.org>
Date:   Fri Oct 31 15:41:34 2014 -0400

    powerpc: use device_online/offline() instead of cpu_up/down()

    In powerpc pseries platform dlpar operations, use device_online() and
    device_offline() instead of cpu_up() and cpu_down().

    Calling cpu_up/down() directly does not update the cpu device offline
    field, which is used to online/offline a cpu from sysfs. Calling
    device_online/offline() instead keeps the sysfs cpu online value
    correct. The hotplug lock, which is required to be held when calling
    device_online/offline(), is already held when dlpar_online/offline_cpu()
    are called, since they are called only from cpu_probe|release_store().

    This patch fixes errors on phyp (PowerVM) systems that have cpu(s)
    added/removed using dlpar operations; without this patch, the
    /sys/devices/system/cpu/cpuN/online nodes do not correctly show the
    online state of added/removed cpus.

    Signed-off-by: Dan Streetman <ddstreet@ieee.org>
    Cc: Nathan Fontenot <nfont@linux.vnet.ibm.com>
    Fixes: 0902a9044fa5 ("Driver core: Use generic offline/online for CPU offline/online")
    Signed-off-by: Michael Ellerman <mpe@ellerman.id.au>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/platforms/pseries/dlpar.c b/arch/powerpc/platforms/pseries/dlpar.c
index 8fbe73d..e1c55bb 100644
--- a/arch/powerpc/platforms/pseries/dlpar.c
+++ b/arch/powerpc/platforms/pseries/dlpar.c
@@ -381,7 +381,7 @@ static int dlpar_online_cpu(struct device_node *dn)
    BUG_ON(get_cpu_current_state(cpu)
      != CPU_STATE_OFFLINE);
    cpu_maps_update_done();
-   rc = cpu_up(cpu);
+   rc = device_online(get_cpu_device(cpu));
    if (rc)
     goto out;
    cpu_maps_update_begin();
@@ -466,7 +466,7 @@ static int dlpar_offline_cpu(struct device_node *dn)
    if (get_cpu_current_state(cpu) == CPU_STATE_ONLINE) {
     set_preferred_offline_state(cpu, CPU_STATE_OFFLINE);
     cpu_maps_update_done();
-    rc = cpu_down(cpu);
+    rc = device_offline(get_cpu_device(cpu));
     if (rc)
      goto out;
     cpu_maps_update_begin();
-- 
1.7.1